File: android/app/src/main/java/dev/wispers/access/android/proxy/SessionManager.kt
package dev.wispers.access.android.proxy

import dev.wispers.access.android.storage.ShareId
import dev.wispers.access.android.storage.ShareRepository
import dev.wispers.access.android.storage.restoreOrInitNode
import dev.wispers.connect.handles.Node
import dev.wispers.connect.handles.QuicConnection
import dev.wispers.connect.handles.QuicStream
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.coroutines.coroutineContext
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout

/**
 * Per-share Node + QuicConnection cache. Equivalent of waclient's `StreamFactory`.
 *
 * Connections are established lazily on first stream request and reused across requests.
 * Stream opens get one retry — if the cached connection has died, we drop it from the
 * pool and reconnect on the second attempt. Opening a stream on a dead connection can
 * still succeed (it's local bookkeeping), so callers must report mid-request failures
 * via [invalidate] — otherwise the dead connection stays cached and every later request
 * fails the same way.
 */
@Singleton
class SessionManager @Inject constructor(
    private val repo: ShareRepository,
) {
    private val nodeMutex = Mutex()
    private val nodes = mutableMapOf<ShareId, Node>()

    private val connMutex = Mutex()
    private val connections = mutableMapOf<ShareId, QuicConnection>()

    /** A stream plus the connection it came from, so [invalidate] can evict exactly that connection. */
    class StreamLease internal constructor(
        val stream: QuicStream,
        internal val connection: QuicConnection,
    )

    // Connect/open can hang silently (e.g. the network changed under an idle
    // connection and packets blackhole), so each attempt gets a deadline.
    suspend fun openStream(shareId: ShareId): StreamLease = try {
        withTimeout(OPEN_TIMEOUT_MS) { tryOpen(shareId) }
    } catch (e: Exception) {
        coroutineContext.ensureActive()
        withTimeout(OPEN_TIMEOUT_MS) { tryOpen(shareId) }
    }

    /** Evicts and closes [lease]'s connection (unless already replaced by a newer one). */
    suspend fun invalidate(shareId: ShareId, lease: StreamLease) {
        evict(shareId, lease.connection)
    }

    /**
     * Drops and closes every cached connection. Called when the device's default
     * network changes: connections on the old network blackhole silently, so
     * reconnecting eagerly beats waiting for per-request timeouts to flush them.
     *
     * @return the affected share ids, for an optional [prewarm].
     */
    suspend fun evictAll(): List<ShareId> {
        val stale = connMutex.withLock {
            val all = connections.toMap()
            connections.clear()
            all
        }
        for (conn in stale.values) runCatching { conn.closeAsync() }
        return stale.keys.toList()
    }

    /**
     * Re-establishes connections for [shareIds] ahead of traffic, so the next
     * request doesn't pay the connection setup (ICE is the expensive part).
     * Failures are swallowed — the per-request path retries lazily anyway.
     */
    suspend fun prewarm(shareIds: List<ShareId>) {
        for (id in shareIds) {
            runCatching { getConnection(id, getNode(id)) }
        }
    }

    /**
     * Tears down [shareId]'s session and deregisters the node from the hub,
     * revoking this device from the share's roster. Deregistration is
     * best-effort: if the hub is unreachable, local removal proceeds anyway
     * and the roster entry lingers until pruned elsewhere.
     */
    suspend fun removeShare(shareId: ShareId) {
        connMutex.withLock { connections.remove(shareId) }
            ?.let { runCatching { it.closeAsync() } }
        val node = nodeMutex.withLock { nodes.remove(shareId) }
            ?: runCatching { repo.storageFor(shareId).restoreOrInitNode().first }.getOrNull()
        node?.let { runCatching { it.logout() } }
    }

    /**
     * Hub-reported availability of the share's serving node, or null when the
     * hub can't be reached (status unknown — typically this device is offline).
     */
    suspend fun isServingNodeOnline(shareId: ShareId): Boolean? = try {
        getNode(shareId).groupInfo()
            .nodes.firstOrNull { it.nodeNumber == SERVING_NODE_NUMBER }
            ?.isOnline
    } catch (_: Exception) {
        null
    }

    private suspend fun tryOpen(shareId: ShareId): StreamLease {
        val node = getNode(shareId)
        val conn = getConnection(shareId, node)
        return try {
            StreamLease(conn.openStream(), conn)
        } catch (e: Exception) {
            evict(shareId, conn)
            throw e
        }
    }

    // NonCancellable: eviction is triggered from catch blocks that may themselves be
    // running cancellation (open timeout) — it must complete anyway, or the dead
    // connection stays cached and every later request hits it.
    private suspend fun evict(shareId: ShareId, conn: QuicConnection) =
        withContext(NonCancellable) {
            connMutex.withLock {
                if (connections[shareId] === conn) connections.remove(shareId)
            }
            runCatching { conn.closeAsync() }
        }

    private suspend fun getNode(shareId: ShareId): Node = nodeMutex.withLock {
        nodes.getOrPut(shareId) {
            val storage = repo.storageFor(shareId)
            val (node, _) = storage.restoreOrInitNode()
            node
        }
    }

    private suspend fun getConnection(shareId: ShareId, node: Node): QuicConnection {
        var established = false
        val conn = connMutex.withLock {
            connections.getOrPut(shareId) {
                established = true
                node.connectQuic(SERVING_NODE_NUMBER)
            }
        }
        // A cache miss means we just reached the serving node — record the
        // contact (cache hits reuse a live connection, so they aren't "new").
        // Done outside the lock to keep the DB write off the connect path.
        if (established) runCatching { repo.markConnected(shareId) }
        return conn
    }

    private companion object {
        const val SERVING_NODE_NUMBER = 1
        const val OPEN_TIMEOUT_MS = 10_000L
    }
}

File: ios/Wispers Access/Tunnel/SessionManager.swift
import Foundation
import WispersConnect

/// A QUIC stream plus the connection it came from, so a mid-request failure can
/// evict exactly that connection.
struct StreamLease: Sendable {
    let stream: QuicStream
    let connection: QuicConnection
}

/// Per-share `Node` + `QuicConnection` cache — the iOS port of the Android
/// `SessionManager`. Connections are established lazily and reused across
/// requests; a stream open gets one retry so a silently-dead connection is
/// dropped and re-established. Callers MUST report mid-request failures via
/// `invalidate` (opening a stream on a dead connection can still succeed as
/// local bookkeeping, so the corpse otherwise stays cached and every later
/// request fails the same way).
actor SessionManager {
    private let storageProvider: @Sendable (ShareID) async throws -> NodeStorage
    private let onConnected: @Sendable (ShareID) async -> Void

    private var nodes: [ShareID: Node] = [:]
    private var connections: [ShareID: QuicConnection] = [:]

    init(
        storageProvider: @escaping @Sendable (ShareID) async throws -> NodeStorage,
        onConnected: @escaping @Sendable (ShareID) async -> Void = { _ in }
    ) {
        self.storageProvider = storageProvider
        self.onConnected = onConnected
    }

    /// Opens a fresh stream to the share's serving node, reusing the cached
    /// connection. One retry: if the first attempt fails, the connection is
    /// evicted and a second attempt reconnects.
    func openStream(_ shareID: ShareID) async throws -> StreamLease {
        do {
            return try await withDeadline(seconds: Self.openTimeout) { try await self.tryOpen(shareID) }
        } catch {
            try Task.checkCancellation()
            return try await withDeadline(seconds: Self.openTimeout) { try await self.tryOpen(shareID) }
        }
    }

    /// Hub-reported availability of the share's serving node (node 1), or nil
    /// when the hub can't be reached (typically this device is offline).
    func isServingNodeOnline(_ shareID: ShareID) async -> Bool? {
        do {
            let node = try await resolveNode(shareID)
            let info = try await node.groupInfo()
            return info.nodes.first { $0.nodeNumber == Self.servingNodeNumber }?.isOnline
        } catch {
            return nil
        }
    }

    /// Evicts and closes `lease`'s connection (unless already replaced).
    func invalidate(_ shareID: ShareID, _ lease: StreamLease) async {
        await evict(shareID, lease.connection)
    }

    /// Drops and closes every cached connection (e.g. on a network change),
    /// returning the affected shares.
    @discardableResult
    func evictAll() async -> [ShareID] {
        let stale = connections
        connections.removeAll()
        for conn in stale.values { try? await conn.closeGracefully() }
        return Array(stale.keys)
    }

    /// Logs the share's node out of the hub — deregistering this device from the
    /// roster — then drops its cached node + connection. Best-effort: an
    /// unreachable hub doesn't block teardown (the roster entry lingers until
    /// pruned). Restores a node from storage if one isn't cached, since `logout`
    /// needs a live handle.
    func logoutAndDiscard(_ shareID: ShareID) async {
        if let conn = connections.removeValue(forKey: shareID) {
            try? await conn.closeGracefully()
        }
        var node = nodes.removeValue(forKey: shareID)
        if node == nil,
            let storage = try? await storageProvider(shareID),
            let restored = try? await storage.restoreOrInit() {
            node = restored.0
        }
        if let node {
            try? await node.logout()
        }
    }

    private func tryOpen(_ shareID: ShareID) async throws -> StreamLease {
        let node = try await resolveNode(shareID)
        let conn = try await resolveConnection(shareID, node: node)
        do {
            return StreamLease(stream: try await conn.openStream(), connection: conn)
        } catch {
            await evict(shareID, conn)
            throw error
        }
    }

    private func resolveNode(_ shareID: ShareID) async throws -> Node {
        if let node = nodes[shareID] { return node }
        // storageProvider applies `overrideHubAddr` before returning, so the
        // node restores against the share's own (possibly self-hosted) hub.
        let storage = try await storageProvider(shareID)
        let (node, _) = try await storage.restoreOrInit()
        nodes[shareID] = node
        return node
    }

    private func resolveConnection(_ shareID: ShareID, node: Node) async throws -> QuicConnection {
        if let conn = connections[shareID] { return conn }
        let conn = try await node.connectQuic(peerNodeNumber: Self.servingNodeNumber)
        connections[shareID] = conn
        await onConnected(shareID)
        return conn
    }

    private func evict(_ shareID: ShareID, _ conn: QuicConnection) async {
        if connections[shareID] === conn { connections.removeValue(forKey: shareID) }
        try? await conn.closeGracefully()
    }

    private static let servingNodeNumber: Int32 = 1
    private static let openTimeout: Double = 10
}
